system/Common.php

```diff
@@ -1046,10 +1045,45 @@ function service(string $name, ...$params)
 	 */
 	function single_service(string $name, ...$params)
 	{
-		// Ensure it's NOT a shared instance
-		array_push($params, false);
+		list($found, $service) = Services::serviceExists($name);
 
-		return Services::$name(...$params);
+		if (! $found)
+		{
+			// The service is not defined anywhere so just return.
+			return null;
+		}
+
+		$method = new ReflectionMethod($service, $name);
+		$count  = $method->getNumberOfParameters();
+		$mParam = $method->getParameters();
+		$params = $params ?? [];
+
+		if ($count === 1)
+		{
+			// This service needs only one argument, which is the shared
+			// instance flag, so let's wrap up and pass false here.
+			return $service::$name(false);
+		}
+
+		// Fill in the params with the defaults, but stop before the last
+		for ($startIndex = count($params); $startIndex <= $count - 2; $startIndex++)
+		{
+			$params[$startIndex] = $mParam[$startIndex]->getDefaultValue();
+		}
+
+		// Ensure the last argument will not create a shared instance
+		$finalCount = count($params);
+
+		if ($finalCount === $count)
+		{
+			$params[$finalCount - 1] = false;
+		}
+		else
+		{
+			$params[] = false;
+		}
```

Comment:
Do we really need this check? 
`$count` already holds the total number of method parameters so I think we can just `$params[$Count - 1] = false;`.

Reply:
The `$finalCount` holds the count of `$params` after it has been filled by ReflectionParameter. My original code for that was simply push in the last value as `false`, but realized what if a user defined **all** the params including the last one? This check will ensure that if the last param was passed as true or any other else, it will be turned to false for no shared instance.

Reply:
But even if I pass in all arguments including `getShared` as true , then `$params[$count - 1] = false;` will still replace this value or am I overseeing something?
Lets take `single_service('exceptions', null, null, null, true);`:
`count($params)` will be `4`
`$count` will be `4` 
Hence we can directly set `$params[3]` to `false`.

Reply:
Consider this example:

`single_service('curlrequest');`
CurlRequest has 4 defined parameters.

The `$params` count after filling up is 3 which is `[[], null, null]`. Then the last param for shared instance will be pushed.

`single_service('curlrequest', [], null, null, true);`

The `$params` count will be 4 and there would be no filling up since the user has defined the params. But we need to change the last param to false to ensure no shared instance. So `$params[4 - 1] = false;`

Reply:
Yes, but still `$count  = $method->getNumberOfParameters();` in both cases holds the value `4` and can directly be used I think.

Reply:
OK. I get what you meant. The extra variable can be dispensed. Thanks.

Reply:
Thats still not what I meant :D

After the for loop ( line 1069)
  `for ($startIndex = count($params); $startIndex <= $count - 2; $startIndex++)`
there is only two possible outcomes. Either, `$params` now hat the size `$count` (i.e., if the user also passed in the `$getShared` argument) or  `$params` now has the size `$count - 1` (because the last loop value for `$startIndex` is `$count - 2`).

In both cases, we can just do `$params[$count - 1] = false;` without further check.
In the first case, this will overwrite the `$getShared` value passed by the user. In the second case, this will create the index for `$getShared` and set it to false.

Both have the same outcome.

So, suggestion is: Remove lines 1075-1076 and 1078-1082.

I hope I was more precise this time :D

Reply:
Ok. That's clever. I didn't notice that. Changing.

Reply:
One more thing: 
In the current implementation, if someone mistakenly passes too many arguments (lets say 5 arguments for 'curlrequest), then on line 1075, `count($params)` will be 5 and `$count` will be 4, so the else clause will be called and create a sixth argument.
This is also solved by the suggestion above.
Btw.: As we are already using reflection here, we might want to check if the user overloads the service, like check if `count($params)` is bigger than `$count` at the beginning and throw an exception if so.

Reply:
I was thinking earlier of how to check for many args passed and that surely solves it.

I don't know for the exception for param overloading as php itself is not throwing an error. But for best practice maybe we can do that. I'll take a second look on it tomorrow.